src/main/java/bitflow4j/steps/fork/distribute/AbstractTagFork.java

```diff
@@ -0,0 +1,129 @@
+package bitflow4j.steps.fork.distribute;
+
+import bitflow4j.Pipeline;
+import bitflow4j.Sample;
+import bitflow4j.misc.Pair;
+import bitflow4j.script.registry.Description;
+import bitflow4j.steps.fork.ScriptableDistributor;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.regex.Pattern;
+import java.util.regex.PatternSyntaxException;
+import java.util.stream.Collectors;
+
+@Description("A Fork that forks the samples based on the provided template. Samples with the same template values of the provided" +
+        " template name will be passed to the same fork. Use ${tagName} Syntax to evaluate the given tags value.")
+public abstract class AbstractTagFork implements ScriptableDistributor {
+
+    private static final Logger logger = Logger.getLogger(AbstractTagFork.class.getName());
+
+    protected final boolean wildcardMatch, regexMatch;
+    private final Map<String, Pipeline> subPipelines = new HashMap<>();
+    private final Map<String, Collection<Pair<String, Pipeline>>> tagValueCache = new HashMap<>(); // Additional cache just for reducing object creation
+    private final Map<String, Pattern> patterns = new HashMap<>();
+
+    protected  Collection<Pair<String, PipelineBuilder>> subPipelineBuilders;
+    private List<String> availableKeys;
+    private Collection<Object> formattedSubPipelines;
+
```

Comment:
No consecutive empty lines please... It is possible to configure most IDEs to format code automatically ;)